espnet/nets/pytorch_backend/e2e_asr.py

```diff
@@ -375,9 +371,103 @@ def calculate_all_attentions(self, xs_pad, ilens, ys_pad):
         """
         with torch.no_grad():
             # encoder
-            hpad, hlens = self.enc(xs_pad, ilens)
+            hpad, hlens, _ = self.enc(xs_pad, ilens)
 
             # decoder
             att_ws = self.dec.calculate_all_attentions(hpad, hlens, ys_pad)
 
         return att_ws
+
+    def subsample_frames(self, x):
+        # subsample frame
+        x = x[::self.subsample[0], :]
+        ilen = [x.shape[0]]
+        h = to_device(self, torch.from_numpy(
+            np.array(x, dtype=np.float32)))
+        h.contiguous()
+        return h, ilen
+
+
+class StreamingE2E(object):
+    """Convenience wrapper over E2E class for streaming recognitions.
+
+    Not recommended for GPUs.
+    """
+
+    def __init__(self, e2e, recog_args, char_list, rnnlm=None):
+        """StreamingE2E constructor.
+
+        :param E2E e2e: E2E ASR object
+        :param recog_args: arguments for "recognize" method of E2E
+        """
+        self._e2e = e2e
+        self._recog_args = recog_args
+        self._char_list = char_list
+        self._rnnlm = rnnlm
+
+        self._e2e.eval()
+
+        self._offset = 0
+        self._previous_encoder_recurrent_state = None
+        self._encoder_states = []
+        self._ctc_posteriors = []
+        self._last_recognition = None
+
+        assert self._recog_args.ctc_weight > 0.0, "StreamingE2E works only with combined CTC and attetion decoders."
+
+    def accept_input(self, x):
+        """Call this method each time a new batch of input is available."""
+
+        h, ilen = self._e2e.subsample_frames(x)
+
+        # Streaming encoder
+        h, _, self._previous_encoder_recurrent_state = self._e2e.enc(
+            h.unsqueeze(0),
+            ilen,
+            self._previous_encoder_recurrent_state
+        )
+        self._encoder_states.append(h.squeeze(0))
+
+        # CTC posteriors for the incoming audio
+        self._ctc_posteriors.append(self._e2e.ctc.log_softmax(h).squeeze(0))
+
+    def _input_window_for_decoder(self, use_all=False):
+        if use_all:
+            return torch.cat(self._encoder_states, dim=0), torch.cat(self._ctc_posteriors, dim=0)
+
+        def select_unprocessed_windows(window_tensors):
+            last_offset = self._offset
```

Comment:
Is it necessary to update `self._offset?` at the end of this function?

Reply:
Should `self._encoder_states` and `self._ctc_posteriors` used for decoding be removed?
Or why don’t you prepare `self.reset()` to clear the stored them?

Reply:
This is the next step - I want to submit (at least a prototype of) streaming attention decoder in a follow-up PR. Then I'll add the necessary operations on the offset, or just clean up the "old" states and posteriors as suggested.